core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -2953,6 +2979,166 @@ private void scheduleAsyncPersistenceInternal(LockedInodePath inodePath) throws
+            for (Inode<?> child : inodeDir.getChildren()) {
+              inodeChildren.put(child.getName(), child);
+            }
+
+            // Iterate over ufs children and process children which do not exist in Alluxio.
+            for (Map.Entry<String, String> ufsEntry : ufsChildren.entrySet()) {
+              if (!inodeChildren.containsKey(ufsEntry.getKey()) && !PathUtils
+                  .isTemporaryFileName(ufsEntry.getKey())) {
+                // Ufs child exists, but Alluxio child does not. Must load metadata.
+                loadMetadata = true;
+                break;
+              }
+            }
+
+            // Iterate over Alluxio children and process persisted children.
+            for (Map.Entry<String, Inode<?>> inodeEntry : inodeChildren.entrySet()) {
+              if (!inodeEntry.getValue().isPersisted()) {
+                // Ignore non-persisted inodes.
+                continue;
+              }
+
+              String ufsFingerprint = ufsChildren.get(inodeEntry.getKey());
+              boolean deleteChild =
+                  !UfsSyncUtils.inodeUfsMatch(inodeEntry.getValue(), ufsFingerprint);
+
+              if (deleteChild) {
+                TempInodePathForDescendant tempInodePath =
+                    new TempInodePathForDescendant(inodePath);
+                tempInodePath.setDescendant(inodeEntry.getValue(),
+                    inodePath.getUri().join(inodeEntry.getKey()));
+
+                try {
+                  deleteInternal(tempInodePath, true, System.currentTimeMillis(), syncDeleteOptions,
+                      journalContext);
+                } catch (DirectoryNotEmptyException | IOException e) {
+                  // Should not happen, since it is an unchecked delete.
+                }
+                // Must load metadata afterwards.
+                loadMetadata = true;
+              }
+            }
+          }
+        }
+      }
+      auditContext.setSucceeded(true);
+    }
+
+    if (loadMetadata) {
+      try (JournalContext journalContext = createJournalContext();
+           LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.READ);
+           FileSystemMasterAuditContext auditContext =
+               createAuditContext("syncMetadata", path, null, inodePath.getInodeOrNull())) {
+        try {
+          mPermissionChecker.checkPermission(Mode.Bits.READ, inodePath);
+        } catch (AccessControlException e) {
+          auditContext.setAllowed(false);
+          throw e;
+        }
+
+        try {
+          loadMetadataAndJournal(inodePath,
+              LoadMetadataOptions.defaults().setCreateAncestors(true).setLoadDirectChildren(true),
+              journalContext);
+        } catch (Exception e) {
+          LOG.debug("Failed to load metadata for path: {}", inodePath.getUri());
+        }
+        auditContext.setSucceeded(true);
+      }
+    }
+
+    mUfsSyncPathCache.addSyncPath(path.getPath());
```

Comment:
this will usually be an update instead of an add - consider renaming to `pathSynced` or `notifyPathSynced`